Enforce specific metadata tags for form/case submissions made via the Submission API

minhaminha opened the pull request:
## Product Description
This PR adds a feature flag that if NOT enabled, blocks form/case submissions made via the Submission API that do not have certain required tags. Saving forms without these tags have shown to disrupt or break certain workflows such as downloading case history. 

The current plan is to enable this feature flag on domains that are known to submit forms without metadata/required tags while giving them time to update their workflows to align with these new changes. 

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Jira Ticket](https://dimagi-dev.atlassian.net/browse/SAAS-13569)
[Original support ticket](https://dimagi-dev.atlassian.net/browse/SUPPORT-13485)

The primary reason this issue was brought up was because case submissions that were being made without the entire metadata node was breaking the ability to download that particular case's case history. Now all form/case submissions will be rejected without a metadata node. Additionally, the `userID`, `instanceID` and `timeEnd` were identified through investigation to be important tags to enforce within the metadata node since all three are required for a specific form submission to be accessible in the Submit History page. `instanceID` in particular is necessary to access the Form properties of a particular submission.

Through further investigation, it's been found that sending a case submission with an empty create or update node also causes a 500 error when attempting to download a case's case history. In keeping with the official [Submission API docs](https://confluence.dimagi.com/display/commcarepublic/Submission+API), this PR also enforces all 3 tags within the case create node. Specifically omitting the `case_name` tag also hides the submission in Submit History. Empty create and update nodes will also be rejected.

If a submission is missing any of these enforced nodes/tags a corresponding error message will let the user know what they're missing. 

This PR also undoes a previous PR which put datadog metric trackers in place to identify which domains were sending bad submissions. 

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

New FF:  [allow_submission_without_metadata](https://staging.commcarehq.org/hq/flags/edit/allow_submission_without_metadata/)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

This PR will not impact existing data, only handling whether new data gets to be saved or not. 

Tested locally and on staging. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

I'll be adding tests soon. 

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

I'll request for QA once this PR has been reviewed and the changes properly communicated to them beforehand, since this will likely break some of the automated workflows they have in place. 

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

minhaminha committed:
Add feature flag to block api submissions without required tags in the metadata + raise and error when it does encounter missing tags

snopoke commented on corehq/form_processor/submission_post.py:
```diff
@@ -500,14 +497,30 @@ def index_case_search(instance, case_models):
+        # INSERT "NOT" HERE when done w testing
+        if ALLOW_SUBMISSION_WITHOUT_METADATA.enabled(domain):
+            form_data = instance.form_data
+            if 'case' in form_data:
+                if 'create' in form_data['case']:
+                    for tag in required_create_tags:
+                        if tag not in form_data['case']['create']:
+                            raise MetadataError('One or more required tags missing in the create node')
+                elif 'update' in form_data['case'] and not isinstance(form_data['case']['update'], dict):
+                    raise MetadataError('Missing nodes in the update node.')
```
This is only handling a case in the root of the data node but cases can appear anywhere in the XML.

I would recommend using the existing tools for extracting case blocks [casexml.apps.case.xform.extract_case_blocks](https://github.com/dimagi/commcare-hq/blob/3b2d4586b18ff87f5b1eed26d28350a9392d5cf8/corehq/ex-submodules/casexml/apps/case/xform.py#L181-L202) or else pushing this down into the case processing layer e.g. adding validation to [CaseCreateAction](https://github.com/dimagi/commcare-hq/blob/cba51a39deebb72248c12c1597ec1b9951cefd8a/corehq/ex-submodules/casexml/apps/case/xml/parser.py#L139-L140)

minhaminha replied:
thanks for pointing this out - I had no idea something like this existed!

snopoke commented on corehq/form_processor/submission_post.py:
```diff
@@ -500,14 +497,30 @@ def index_case_search(instance, case_models):
+                            raise MetadataError('One or more required tags missing in the create node')
+                elif 'update' in form_data['case'] and not isinstance(form_data['case']['update'], dict):
+                    raise MetadataError('Missing nodes in the update node.')
+            if instance.metadata:
+                for attr in required_metadata_tags:
+                    if getattr(instance.metadata, attr) is None:
+                        raise MetadataError('One or more required tags missing in the metadata node')
+            else:
+                raise MetadataError('Missing metadata node.')
```
At least this (potentially the whole validation block) should be moved out of the case processing function and placed somewhere before we call get_locked_forms so that we can avoid doing the extra work involved in getting locks etc.

snopoke commented on corehq/form_processor/submission_post.py:
```diff
@@ -328,7 +320,12 @@ def run(self):
-                        case_stock_result = self.process_xforms_for_cases(xforms, case_db, self.timing_context)
+                        case_stock_result = self.process_xforms_for_cases(xforms, case_db,
+                                                                          self.domain, self.timing_context)
+                    except MetadataError as e:
+                        instance.problem = e
+                        instance.state = XFormInstance.ERROR
+                        submission_type = 'error'
```
This should be incorporated into the 'known errors' block below which will give you the standard error handling (you are currently not saving the form and not responding with the correct OpenRosa error nature).

minhaminha replied:
I'm reluctant to use `_handle_known_error` since the whole point of this validation is to prevent forms with missing tags/nodes from being saved in the first place, which is what seems to be doing, regardless of what the error is. Though I may be misunderstanding what it does since I assume we also don't want forms with other exceptions (IllegalCaseId, UsesReferrals, MissingProductId, etc.) to be saved, but we do anyways by calling _handle_known_error??

snopoke reviewed (commented):
Needs tests (which I see you mention you'll be adding soon 👍🏻)

minhaminha committed:
Use extract_case_blocks + validate submission before locking form

minhaminha commented:
This PR is unfortunately going to be put on hold until there's enough time to work on the rollout of these changes, which is primarily notifying the handful of domains we've identified will be affected by this change. So in the interest of time and because its likely this might get picked up by someone else, I'm going to work on addressing reviewers' comments until its at a state where all that's left to do is add tests and have QA test it. 

Thanks @snopoke for being so quick to provide feedback!

orangejenny committed:
Merge branch 'master' into ml/enforce-metadata-ff
